File: Python/NLP/RestuarantReviews-NLP.py
# Natural Language Processing 

# Import Libraries 
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt 

# Rad Dateset 
# Use Quoting to ignore the double quotation in data 
df = pd.read_csv(r"C:\Users\snapp\Data-Projects\Datasets\Restaurant_Reviews.tsv", delimiter = '\t', quoting = 3)

# Cleaning the texts 
import re
import nltk
nltk.download('stopwords')
from nltk.corpus import stopwords 
from nltk.stem.porter import PorterStemmer

corpus = []
for i in range(0,df.shape[0]):
    # Clean the punctuations, numbers, ... except alphabet and spaces
    review = re.sub('[^a-zA-Z]', ' ', df['Review'][i])
    # Change all the capital letter to lower letter
    review = review.lower()
    # Remove irrelevent words - stopwords/Stemming: Taking the root of the words/Changing the text to a list of words 
    review = review.split()
    ps = PorterStemmer()
    # Set is faster than list in python 
    review = [ps.stem(word) for word in review if not word in set(stopwords.words('english'))]
    # Joining the word with space
    review = ' '.join(review)
    # Append the cleaned text to the list 
    corpus.append(review)

# Creating the Bag of words model - create a sparse matrix(create one column for each word)
from sklearn.feature_extraction.text import CountVectorizer
## countvectorizer has a lot of feature for cleaning data but it is better clean data step by step 
cv = CountVectorizer(max_features= 1500)
# Toarray is necessary to convert it to matrix 
X = cv.fit_transform(corpus).toarray()
y = df.iloc[:,-1]

# Classification Model: the best model for NLP: NB, DT, RF 
# Split Data into Train and Test set 
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Feature Scaling: it is not necessary because all the data is: 0,1,2, and few numbers

# Fit Naive Bayes to the Training set 
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(X_train, y_train)

# Predict the Test Set Results 
y_pred = classifier.predict(X_test)

# Confusion Matrix
# The Data is small  
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(y_test, y_pred)

# Accuracy 
from sklearn.metrics import accuracy_score
ac = accuracy_score(y_test, y_pred)

File: R/NLP/RestuarantReviews-NLP.R
# Natural Language Processing 

# Importing the Datasets
dataset = read.delim("C:/Users/snapp/Data-Projects/Datasets/Restaurant_Reviews.tsv", quote = '', stringsAsFactors = FALSE)

# Cleaning the Text 
# install.packages('tm')
# install.packages('SnowballC')
library(tm)
library(SnowballC)
corpus = VCorpus(VectorSource(dataset$Review))
# See the First Review 
review = as.character(corpus[[1]])

# Change all the word from uppercase to lowercase 
corpus = tm_map(corpus, content_transformer(tolower))

# Remove all the numbers 
corpus = tm_map(corpus, removeNumbers)

# Remove all the punctuation 
corpus = tm_map(corpus, removePunctuation)

# Remove Nonrelevant Words like: it, is, this, that, the, ... 
corpus = tm_map(corpus, removeWords, stopwords())

# Stemming: extract root of the word
corpus = tm_map(corpus, stemDocument)

# Remove all the extra spaces 
corpus = tm_map(corpus, stripWhitespace)

# Creating the Bag of Words model 
dtm = DocumentTermMatrix(corpus)

# To Reduce Sparsity we can filter most frequent words 
dtm = removeSparseTerms(dtm, 0.999)

# ML Model that is good for NLP: NB, DT, RF 
df = as.data.frame(as.matrix(dtm))
df$Liked = dataset$Liked

# Encoding the Target Feature aas Factor 
df$Liked = factor(df$Liked, levels = c(0,1))

# Split Dataset into train test split 
library(caTools)
# Set random state for seeing same data 
set.seed(123)

split = sample.split(df$Liked, SplitRatio = 0.8)
train_set = subset(df, split == TRUE)
test_set = subset(df, split == FALSE)

# Feature Scaling is not necessary 
# Fit Random Forest To the training set and Predict the test set results  
# install.packages('randomForest')
library(randomForest)
classifier = randomForest(x = train_set[-692],
                          y =train_set$Liked,
                          ntree = 10)

# Predict Test Results 
y_pred = predict(classifier, newdata = test_set[-692])

# Make Confusion Matrix 
CM = table(test_set[,692], y_pred)
print(CM)



